File: paperless/types/formDefinition.ts
/**
 * TypeScript type definitions mirroring the C# FormDefinition model.
 * Used for round-trip Upload Excel / Export Excel workflow.
 */

export interface FormDefinition {
  workbook: WorkbookMetadata;
  sheets: SheetDefinition[];
  clusters: ClusterDefinition[];
  images: ImageDefinition[];
  metadata: Record<string, string>;
}

export interface WorkbookMetadata {
  title: string;
  author: string;
  created: string;
  modified: string;
  version: string;
  description: string;
}

export interface SheetDefinition {
  id: string;
  name: string;
  index: number;
  pageSettings: PageSettings;
  printArea: PrintAreaInfo | null;
  backgroundImage: string | null;
  backgroundWidth: number;
  backgroundHeight: number;
  thumbnail: string | null;
  rowHeights: Record<number, number>;
  columnWidths: Record<number, number>;
  mergedCells: MergedCellInfo[];
  freezePane: string | null;
  cellStyles: Record<string, CellStyleInfo>;
  cellValues: Record<string, string>;
  metadata: Record<string, string>;
}

export interface PageSettings {
  paperSize: string;
  orientation: string;
  widthPt: number;
  heightPt: number;
  leftMargin: number;
  topMargin: number;
  rightMargin: number;
  bottomMargin: number;
  centerHorizontally: boolean;
  centerVertically: boolean;
  zoom: number;
  fitToPagesWide: number;
  fitToPagesTall: number;
}

export interface PrintAreaInfo {
  address: string;
  leftPt: number;
  topPt: number;
  widthPt: number;
  heightPt: number;
  cols: number;
  rows: number;
}

export interface MergedCellInfo {
  address: string;
  cellAddress: string;
  leftPt: number;
  topPt: number;
  widthPt: number;
  heightPt: number;
}

export interface CellStyleInfo {
  fontName: string | null;
  fontSize: number | null;
  bold: boolean | null;
  italic: boolean | null;
  underline: boolean | null;
  color: string | null;
  fillColor: string | null;
  borderTop: string | null;
  borderBottom: string | null;
  borderLeft: string | null;
  borderRight: string | null;
  horizontalAlignment: string | null;
  verticalAlignment: string | null;
  wrapText: boolean | null;
}

export interface ClusterDefinition {
  clusterId: string;
  name: string;
  type: string;
  sheetId: string;
  cellAddress: string;
  left: number;
  right: number;
  top: number;
  bottom: number;
  leftPt: number;
  topPt: number;
  widthPt: number;
  heightPt: number;
  inputParameters: Record<string, string>;
  visibility: string;
  readonly: boolean;
  remarks: string;
  functions: string[];
  metadata: Record<string, string>;
}

export interface ImageDefinition {
  id: string;
  sheetId: string;
  name: string;
  leftPt: number;
  topPt: number;
  widthPt: number;
  heightPt: number;
  data: string | null;
  format: string;
}

// ── API Response Types ──

export interface UploadExcelResponse {
  success: boolean;
  message: string;
  data: {
    formDefinition: FormDefinition;
    fieldCount: number;
    sheetCount: number;
    templateId: string;
    workbookDownloadUrl: string;
  };
}

export interface OutputExcelResponse {
  success: boolean;
  message: string;
  data: {
    workbookPath: string;
    downloadUrl: string;
    fieldCount: number;
    sheetCount: number;
  };
}

File: ExcelAPI/ExcelAPI/Models/FormDefinition.cs
namespace ExcelAPI.Models
{
    public class FormDefinition
    {
        public WorkbookMetadata Workbook { get; set; } = new();
        public List<SheetDefinition> Sheets { get; set; } = new();
        public List<ClusterDefinition> Clusters { get; set; } = new();
        public PageSettings? FieldsPageSettings { get; set; }
        public List<ImageDefinition> Images { get; set; } = new();
        public Dictionary<string, string> Metadata { get; set; } = new();
    }

    public class WorkbookMetadata
    {
        public string Title { get; set; } = "Untitled Form";
        public string Author { get; set; } = "";
        public string Created { get; set; } = "";
        public string Modified { get; set; } = "";
        public string Version { get; set; } = "1.0";
        public string Description { get; set; } = "";
    }

    public class SheetDefinition
    {
        public string Id { get; set; } = "";
        public string Name { get; set; } = "";
        public int Index { get; set; }
        public PageSettings PageSettings { get; set; } = new();
        public PrintAreaInfo? PrintArea { get; set; }
        public string? BackgroundImage { get; set; }
        public int BackgroundWidth { get; set; }
        public int BackgroundHeight { get; set; }
        public string? Thumbnail { get; set; }
        public Dictionary<int, double> RowHeights { get; set; } = new();
        public Dictionary<int, double> ColumnWidths { get; set; } = new();
        public List<MergedCellInfo> MergedCells { get; set; } = new();
        public string? FreezePane { get; set; }
        public Dictionary<string, CellStyleInfo> CellStyles { get; set; } = new();
        public Dictionary<string, string> CellValues { get; set; } = new();
        public Dictionary<string, string> Metadata { get; set; } = new();
    }

    public class PageSettings
    {
        public string PaperSize { get; set; } = "Letter";
        public string Orientation { get; set; } = "portrait";
        public double WidthPt { get; set; } = 612;
        public double HeightPt { get; set; } = 792;
        public double LeftMargin { get; set; } = 70;
        public double TopMargin { get; set; } = 70;
        public double RightMargin { get; set; } = 70;
        public double BottomMargin { get; set; } = 70;
        public bool CenterHorizontally { get; set; }
        public bool CenterVertically { get; set; }
        public int Zoom { get; set; } = 100;
        public int FitToPagesWide { get; set; }
        public int FitToPagesTall { get; set; }
    }

    public class PrintAreaInfo
    {
        public string Address { get; set; } = "";
        public double LeftPt { get; set; }
        public double TopPt { get; set; }
        public double WidthPt { get; set; }
        public double HeightPt { get; set; }
        public int Cols { get; set; }
        public int Rows { get; set; }
    }

    public class MergedCellInfo
    {
        public string Address { get; set; } = "";
        public string CellAddress { get; set; } = "";
        public double LeftPt { get; set; }
        public double TopPt { get; set; }
        public double WidthPt { get; set; }
        public double HeightPt { get; set; }
    }

    public class CellStyleInfo
    {
        public string? FontName { get; set; }
        public double? FontSize { get; set; }
        public bool? Bold { get; set; }
        public bool? Italic { get; set; }
        public bool? Underline { get; set; }
        public string? Color { get; set; }
        public string? FillColor { get; set; }
        public string? BorderTop { get; set; }
        public string? BorderBottom { get; set; }
        public string? BorderLeft { get; set; }
        public string? BorderRight { get; set; }
        public string? HorizontalAlignment { get; set; }
        public string? VerticalAlignment { get; set; }
        public bool? WrapText { get; set; }
    }

    public class ClusterDefinition
    {
        public string ClusterId { get; set; } = "";
        public string Name { get; set; } = "";
        public string Type { get; set; } = "";
        public string SheetId { get; set; } = "";
        public string CellAddress { get; set; } = "";
        public double Left { get; set; }
        public double Right { get; set; }
        public double Top { get; set; }
        public double Bottom { get; set; }
        public double LeftPt { get; set; }
        public double TopPt { get; set; }
        public double WidthPt { get; set; }
        public double HeightPt { get; set; }
        public Dictionary<string, string> InputParameters { get; set; } = new();
        public string Visibility { get; set; } = "visible";
        public bool Readonly { get; set; }
        public string Remarks { get; set; } = "";
        public List<string> Functions { get; set; } = new();
        public Dictionary<string, string> Metadata { get; set; } = new();
    }

    public class ImageDefinition
    {
        public string Id { get; set; } = "";
        public string SheetId { get; set; } = "";
        public string Name { get; set; } = "";
        public double LeftPt { get; set; }
        public double TopPt { get; set; }
        public double WidthPt { get; set; }
        public double HeightPt { get; set; }
        public byte[]? Data { get; set; }
        public string Format { get; set; } = "png";
    }
}
